etc/afpd/afpstats.c

```diff
@@ -56,34 +94,78 @@ static gpointer afpstats_thread(gpointer _data _U_)
     ctxt = g_main_context_new();
```

Comment:
Is this running in a non-main thread? If yes, you'll probably need to set the main-context to be the thread-default main context:

```
    ctxt = g_main_context_new();
    g_main_context_push_thread_default(ctxt);

    ... do all the other stuff that happens in this thread ...

    g_main_context_pop_thread_default(ctxt);
    g_main_context_unref(ctxt);
}
```

Otherwise, `handle_method()` will get called from a different thread that is not this one (most likely your main thread). Or is that what you were assuming? I don't know what your multi-threading design in this project looks like.

Reference: https://developer.gnome.org/documentation/tutorials/main-contexts.html

With dbus-glib, this might have worked accidentally because dbus-glib was not thread-safe, so it would have called methods/callbacks on some arbitrary thread that is not the one you expected...

Reply:
It would probably be useful to add some

```
g_debug ("called %s in thread %p", G_STRFUNC, g_thread_self ());
```

(or the equivalent with your `LOG` macro) in strategic places, to check that everything is being called in the thread where you expected it to be called.

Reply:
This is definitely running in a non-main thread!

I'm pleased to report that this was the missing piece that made it work.

```
$ afpstats 
Connected user   PID      Login time        State          Mounted volumes
dmark            573288   Sep 20 10:29:45   active         My Share
```